freqtrade/freqai/freqai_util.py

```diff
@@ -0,0 +1,115 @@
+    )
+
+
+def get_timerange_from_ready_models(models_path: Path) -> Tuple[TimeRange, str, Dict[str, Any]]:
+    """
+    Returns timerange information based on a FreqAI model directory
+    :param models_path: FreqAI model path
+
+    :returns: a Tuple with (backtesting_timerange: Timerange calculated from directory,
+    backtesting_string_timerange: str timerange calculated from
+    directory (format example '20020822-20220830'), \
+    pairs_end_dates: Dict with pair and model end training dates info)
+    """
+    all_models_end_dates = []
+    pairs_end_dates: Dict[str, Any] = get_pairs_timestamps_training_from_ready_models(models_path)
+    for key in pairs_end_dates:
+        for model_end_date in pairs_end_dates[key]:
+            if model_end_date not in all_models_end_dates:
+                all_models_end_dates.append(model_end_date)
+
+    if len(all_models_end_dates) == 0:
+        raise OperationalException(
+            'At least 1 saved model is required to '
+            'run backtest with the freqai-backtest-live-models option'
+        )
+
+    if len(all_models_end_dates) == 1:
+        logger.warning(
+            "Only 1 model was found. Backtesting will run with the "
+            "timerange from the end of the training date to the current date"
+        )
+
+    finish_timestamp = int(datetime.now(tz=timezone.utc).timestamp())
+    if len(all_models_end_dates) > 1:
+        # After last model end date, use the same period from previous model
+        # to finish the backtest
+        all_models_end_dates.sort(reverse=True)
+        finish_timestamp = all_models_end_dates[0] + \
+            (all_models_end_dates[0] - all_models_end_dates[1])
+
+    all_models_end_dates.append(finish_timestamp)
+    all_models_end_dates.sort()
+    start = datetime.fromtimestamp(min(all_models_end_dates), tz=timezone.utc)
+    stop = datetime.fromtimestamp(max(all_models_end_dates), tz=timezone.utc)
+    end_date_string_timerange = stop
+    if (
+        finish_timestamp < int(datetime.now(tz=timezone.utc).timestamp()) and
+        datetime.now(tz=timezone.utc).strftime('%Y%m%d') != stop.strftime('%Y%m%d')
+    ):
+        # add 1 day to string timerange to ensure BT module will load all dataframe data
+        end_date_string_timerange = stop + timedelta(days=1)
+
+    backtesting_string_timerange = (
+        f"{start.strftime('%Y%m%d')}-{end_date_string_timerange.strftime('%Y%m%d')}"
+    )
+    backtesting_timerange = TimeRange(
+        'date', 'date', min(all_models_end_dates), max(all_models_end_dates)
+    )
+    return backtesting_timerange, backtesting_string_timerange, pairs_end_dates
+
+
+def get_pairs_timestamps_training_from_ready_models(models_path: Path) -> Dict[str, Any]:
+    """
+    Scan the models path and returns all pairs end training dates (timestamp)
+    :param models_path: FreqAI model path
+
+    :returns:
+    :pairs_end_dates: Dict with pair and model end training dates info
```

Comment:
it's interesting that people seem to think :returns: is the proper syntax (i've noticed this elsewhere, too)- it's actually :return: <value>.
Freqtrade "mostly" follows this (it's not always noticed in reviews).

Also, the actual variable name you're returning is missleading info - that's not visible to the outside / caller, and if that were renamed, it'll become an outdated fragment very soon.